### DISCUSSION TITLE
Asynchronous APIs and Event Driven Architecture and Event Management

### DISCUSSION THREAD

**Discussion by mathusmith** [tewhatuora/api-standards#147] [General]
The Asynchronous API section of the standard provides good coverage of this important compliment to event driven architectures (EDA).   However without broader coverage of event driven architectures and event management there's the risk of the broader capability and standards being shoehorning into a Asynchronous API mould. 

This being a set of **API Standards**, I don't think it's the place to provide this wider coverage but I'm keen to hear peoples thoughts on this and on how it can be addressed.

**Comment by HG-TWO** [comment]
Since the REST API standards are not specifying the target architecture, I lean toward the approach that the asynchronous api standards should not specify the target architecture. EDA is an architecture style (and a bloody good one) where you could utilise the asynchronous api and standards, but I am sure there are other architectures could make use of asynchronous api's. I noticed that the MQTT is mentioned as a supported protocol in the standards. This could be a good example of the use the asynchronous api: for the purpose of IoT data collection, but not about event driven architecture. I would suggest the approach by saying that EDA is well supported by the API Standards (but is not limiting the use of the standards).

**Comment by jdttaylor** [comment]
I tend to agree. When I think of event driven architecture I think of a different messaging pattern with different protocols. When I hear Async API, I'm not necessarily thinking of a different messaging pattern as I can use the same protocol that I use for my sync API. 

When I hear Async API I think:
* Fire and forget
* long running requests

When I hear Event Driven Architecture I think:
* Guaranteed delivery
* Queues and topics
* Messaging protocols, e.g. JMS, MQTT, etc

If I look at the overall documentation you have a section on Synchronous, many of the principles you have there apply to Aynchronous APIs. Wouldn't it be best to talk about the API protocols you want to support / recommend, e.g. REST, SOAP, JMS, MQTT, etc.

As a developer or an architect I want to know that Im using the right protocol for the challenge I'm trying resolve. I want to know that Im implementing that protocol to the API standard. If I'm unsure what protocol to use to resolve a challenge, I would like to reference the API standards to help guide me.

**Comment by spud023** [comment]
Hi @jdttaylor and @HG-TWO ... I think this is a useful discussion. I came at it through a lower-level approach in [Issue 143](https://github.com/tewhatuora/api-standards/issues/143), where I suggest we introduce some more language about APIs to facilitate the discussion on the more complex roles that occur in the Asynchonous interaction styles, and how the concept of an API in this spec so far is almost always used with a "Producer" or "Consumer" prefix... which works OK for Synchronous blocking calls because the "Producer API" is the server waiting for interactions from the "Consumer API". But in Asych patters such that those discussed in the standard, and referred to in this discussion, the different roles in the pattern all support the "server role" by accepting incoming calls, for example, for event notification messages. I propose we introduce smaller grained concept which could be called a "Granular API" or "Technical API" so that this can be discussed at the level below the API designer, and mention the API roles of all parties in a particular Asynch design pattern. I won't repeat that all here... see the link above.
